core/src/main/java/jenkins/model/JobGlobalBuildDiscarderStrategy.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * The MIT License
+ *
+ * Copyright 2019 Daniel Beck
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package jenkins.model;
+
+import hudson.Extension;
+import hudson.model.Job;
+import org.jenkinsci.Symbol;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+import org.kohsuke.stapler.DataBoundConstructor;
+
+import javax.annotation.Nonnull;
+import java.io.IOException;
+
+/**
+ * Periodically call a job's configured build discarder in the background.
+ */
+@Restricted(NoExternalUse.class)
+public class JobGlobalBuildDiscarderStrategy extends GlobalBuildDiscarderStrategy {
+
+    @DataBoundConstructor
+    public JobGlobalBuildDiscarderStrategy() {
+        // required for data binding
+    }
+
+    @Override
+    public boolean isApplicable(Job<?, ?> job) {
+        return job.getBuildDiscarder() != null;
+    }
+
+    @Override
+    public void apply(Job<?, ?> job) throws IOException, InterruptedException {
+        job.logRotate();
```

Comment:
`Job` is a non-final, public, abstract class, and `logRotate()` is a non-final, public method on `Job` - and neither are annotated with `@NoExternalUse`.

It is not guaranteed that `job.logRotate()` will actually use the build discarder that was checked in `isApplicable(...)`. This may cause the documentation around this step to fail to align with what actually happens.

Is there a reason to prefer `job.logRotate()` over `job.getBuildDiscarder().perform( job )`?

Reply:
Good question. I think the expectation is that `#logRotate()` not be overridden, but it's not enforced, true.

I am unsure which is the more correct option.

`#logRotate()` is called by `Run#execute(RunExecution)`, so an argument can be made that this is the expected way to rotate logs, and if the job overrides it in a way that differs from what `getBuildDiscarder().perform(…)` does, `#logRotate()` should be take precedence.

OTOH, the way this feature is explained on the UI would indicate we want `getBuildDiscarder().perform(…)` instead. Obviously, much of that is by necessity.